pkg/inspector/inspector.go

```diff
@@ -0,0 +1,139 @@
+// Copyright © 2022 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package inspector
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/conduitio/conduit/pkg/foundation/log"
+	"github.com/conduitio/conduit/pkg/record"
+	"github.com/google/uuid"
+)
+
+// Session wraps a channel of records and provides:
+// 1. a way to send records to it asynchronously
+// 2. a way to know if it's closed or not
+type Session struct {
+	C chan *record.Record
+
+	logger  log.CtxLogger
+	onClose func()
+	closed  bool
+}
+
+func (s *Session) Close() {
+	s.onClose()
+	close(s.C)
+	s.closed = true
+}
+
+func (s *Session) IsClosed() bool {
+	return s.closed
+}
+
+// send sends a record to the session's channel asynchronously.
+// By sending the record asynchronously we make sure that even
+// if nothing is getting records from the session that we
+// can still continue working.
+func (s *Session) send(r *record.Record) {
+	c := s.C
+	go func() {
```

Comment:
I'm not too happy with this solution. An async send is needed not to block the inspector (and with that, the connector).

Reply:
**Option 2**( I'm currently looking into) is having buffered channels in sessions. When a channel reaches its full size, new messages will be discarded.

Reply:
Apart from the problem with a ton of goroutines, we are also throwing ordering out of the window. If you spawn two goroutines that try to write to a channel you can't guarantee which one will be first.

When we were talking about this in the design I had the `default` statement in mind:
```
select {
case c <- r:
    // record sent to c
default:
    // c is full
}
```

The channel has to be buffered for this to work though. Maybe that's what you already meant with "Option 2".

Reply:
Oh, I didn't even know you can do that with buffered channels:) 

But yes, that's what I meant with option 2 (and that's what I implemented in the end).